### PR TITLE
PILOT-1533 Search items by attributes

### PR DESCRIPTION
**Pull request by erikdvlp**
## Summary

- Allow searching ElasticSearch `metadata-items` by attribute values.
	- Attributes parameter accepts a JSON object (e.g. `{"key1":["value1","value2","value3"],"key2":"value1"}`).
	- Lists are interpreted as contains / match any. Strings are interpreted as must match exactly.
- Uses the new attributes format from: https://github.com/PilotDataPlatform/metadata-event-handler/pull/13

## JIRA Issues

https://indocconsortium.atlassian.net/browse/PILOT-1533

## Type of Change

- [x] New feature (non-breaking change which adds functionality)

## Testing

Are there any new or updated tests to validate the changes?

- [x] Yes

## Test Directions

Copy and paste attributes JSON objects from Portal API calls and see if they are handled correctly.

### CODE REVIEW

File: search/components/metadata_item/parameters.py [hunk 1]
```diff
@@ -47,8 +51,18 @@ class MetadataItemFilterParameters(FilterParameters):
...
     template_id: UUID | None = Query(default=None)
+    attributes: str | None = Query(default=None)
     is_archived: bool | None = Query(default=None)
 
+    @validator('attributes')
+    def attributes_validation(cls, v):
+        if v:
+            try:
+                json.loads(v)
+            except Exception:
```
Comment (mloshakov) [inline search/components/metadata_item/parameters.py:62]:
```suggestion
            except ValueError:
```

File: search/components/metadata_item/parameters.py [hunk 2]
```diff
@@ -47,8 +51,18 @@ class MetadataItemFilterParameters(FilterParameters):
...
+                raise HTTPException(422, detail=param_validation_error('attributes', 'Must be valid JSON'))
```
Comment (mloshakov) [inline search/components/metadata_item/parameters.py:63]: Here you don't need to raise `HTTPException`, but `ValueError` with the custom message.

File: search/components/metadata_item/filtering.py [hunk 3]
```diff
@@ -35,6 +35,7 @@ class MetadataItemFiltering(Filtering):
     size_gte: int | None = None
     size_lte: int | None = None
     template_id: UUID | None = None
+    attributes: dict | None = None
```
Comment (mloshakov) [inline search/components/metadata_item/filtering.py:38]:
```suggestion
    attributes: dict[str, Any] | None = None
```

File: search/components/metadata_item/models.py [hunk 4]
```diff
@@ -62,7 +62,7 @@ class MetadataItem(BaseModel):
     tags: list[str]
     template_id: UUID | None
     template_name: str
-    attributes: list[MetadataItemAttribute]
+    attributes: dict | None
```
Comment (mloshakov) [inline search/components/metadata_item/models.py:65]:
```suggestion
    attributes: dict[str, Any] | None
```

File: search/components/metadata_item/schemas.py [hunk 5]
```diff
@@ -47,7 +40,7 @@ class MetadataItemSchema(BaseSchema):
     tags: list[str]
     template_id: UUID | None
     template_name: str
-    attributes: list[MetadataItemAttributeSchema]
+    attributes: dict | None
```
Comment (mloshakov) [inline search/components/metadata_item/schemas.py:43]:
```suggestion
    attributes: dict[str, Any] | None
```

File: tests/fixtures/components/metadata_item.py [hunk 6]
```diff
@@ -50,7 +49,7 @@ def generate(  # noqa: C901
         tags: list[str] = ...,
         template_id: UUID | None = ...,
         template_name: str = ...,
-        attributes: list[dict[str, Any]] = ...,
+        attributes: dict | None = ...,
```
Comment (mloshakov) [inline tests/fixtures/components/metadata_item.py:52]:
```suggestion
        attributes: dict[str, Any] | None = ...,
```

File: tests/fixtures/components/metadata_item.py [hunk 7]
```diff
@@ -258,7 +259,7 @@ async def bulk_create(
         tags: list[str] = ...,
         template_id: UUID | None = ...,
         template_name: str = ...,
-        attributes: list[dict[str, Any]] = ...,
+        attributes: dict = ...,
```
Comment (mloshakov) [inline tests/fixtures/components/metadata_item.py:262]:
```suggestion
        attributes: dict[str, Any] | None = ...,
```

File: tests/fixtures/components/metadata_item.py [hunk 8]
```diff
@@ -167,7 +168,7 @@ async def create(
...
```
Comment (mloshakov) [inline tests/fixtures/components/metadata_item.py:171]:
```suggestion
        attributes: dict[str, Any] | None = ...,
```

File: search/components/metadata_item/parameters.py (hunk 2 above)
Comment (erikdvlp) [reply search/components/metadata_item/parameters.py:63]: Usually yes, but Pydantic validators are meant to operate on the JSON body of a POST/PUT/PATCH/DELETE call, not the params of a GET call. If I throw a `ValueError` here, only a 500 with no description will be returned to the client. See [this discussion](https://github.com/tiangolo/fastapi/issues/318) for example.

As a way around it, I'm directing throwing an `HTTPException` so the client gets a 422 and a description.

File: search/components/metadata_item/parameters.py (hunk 1 above)
Comment (erikdvlp) [reply search/components/metadata_item/parameters.py:62]: See `ValueError` thread below.

File: search/components/metadata_item/parameters.py (hunk 2 above)
Comment (mloshakov) [reply search/components/metadata_item/parameters.py:63]: Ah, okay. Still, there is a better way to do it.

See https://github.com/PilotDataPlatform/project/pull/19 by Hugo, specifically this part https://github.com/PilotDataPlatform/project/pull/19/files#diff-1384eb1db34f96576150f8e4e901eb262f08ae190f77fc4d9b3e3c616763ab4cR115

**Review by mloshakov — CHANGES_REQUESTED** [review]
See this comment https://github.com/PilotDataPlatform/search/pull/32#discussion_r957025000

**Review by mloshakov — APPROVED** [review]
👍

**Pull request merged**